api/src/main/java/org/openmrs/module/m2sysbiometrics/service/impl/UpdateServiceImpl.java

```diff
@@ -40,19 +49,34 @@ public void updateLocally(BiometricSubject subject, M2SysCaptureResponse fingerS
     }
 
     @Override
-    public void updateNationally(BiometricSubject nationalSubject, M2SysCaptureResponse fingerScan) {
-        String response = nationalBioServerClient.update(nationalSubject.getSubjectId(), fingerScan.getTemplateData());
-        M2SysResults results = XmlResultUtil.parse(response);
-        if (!results.isUpdateSuccess()) {
-            LOGGER.error("Unable to update fingerprints nationally for: " + nationalSubject.getSubjectId()
-                    + ", reason: " + results.firstValue());
-            handleNationalUpdateError(nationalSubject, fingerScan);
+    public void updateNationally(M2SysCaptureResponse fingerScan) {
+        try {
+            BiometricMatch nationalMatch = searchService.findMostAdequateNationally(fingerScan);
+            if (nationalMatch != null) {
+                BiometricSubject nationalSubject = new BiometricSubject(nationalMatch.getSubjectId());
+
+                String response = nationalBioServerClient.update(nationalSubject.getSubjectId(),
+                        fingerScan.getTemplateData());
+                M2SysResults results = XmlResultUtil.parse(response);
+                if (!results.isUpdateSuccess()) {
+                    throw new M2SysBiometricsException("Unable to update fingerprints nationally for: "
+                            + nationalSubject.getSubjectId()
+                            + ", reason: " + results.firstValue());
+                }
+            } else {
+                registrationService.registerNationally(fingerScan);
+            }
+        } catch (Exception e) {
+            LOGGER.error("Update with the national fingerprint server failed.", e);
+            String subjectId = searchService.findMostAdequateLocally(fingerScan).getSubjectId();
```

Comment:
Couldn't it just be empty String? It's not used in task executions later. You could also simplify handleNationalUpdateError method then.

Reply:
or sth like "not needed" String

Reply:
ok I will correct it with NOT_NEEDED